sycl/source/detail/memory_manager.cpp

```diff
@@ -990,6 +991,164 @@ void MemoryManager::memset_2d_usm(void *DstMem, QueueImplPtr Queue,
       Height, DepEvents.size(), DepEvents.data(), OutEvent);
 }
 
+static void memcpyToDeviceGlobalUSM(QueueImplPtr Queue,
+                                    DeviceGlobalMapEntry *DeviceGlobalEntry,
+                                    size_t NumBytes, size_t Offset,
+                                    const void *Src,
+                                    std::vector<RT::PiEvent> DepEvents,
+                                    RT::PiEvent *OutEvent) {
+  // Zero-initialization is only needed if we do not overwrite the entire
+  // device_global.
+  bool NeedZeroInit =
+      Offset != 0 || NumBytes != DeviceGlobalEntry->MDeviceGlobalTSize;
+  // Get or allocate USM memory for the device_global.
+  DeviceGlobalUSMMem &DeviceGlobalUSM =
+      DeviceGlobalEntry->getOrAllocateDeviceGlobalUSM(
+          Queue,
+          /*ZeroInit=*/NeedZeroInit);
+  void *Dest = DeviceGlobalUSM.getPtr();
+  std::optional<RT::PiEvent> ZIEvent =
+      DeviceGlobalUSM.getZeroInitEvent(Queue->getPlugin());
+
+  // If there is a zero-initializer event the memory operation should wait for
+  // it.
+  if (ZIEvent.has_value())
+    DepEvents.push_back(*ZIEvent);
+
+  MemoryManager::copy_usm(Src, Queue, NumBytes,
+                          reinterpret_cast<char *>(Dest) + Offset, DepEvents,
```

Comment:
I hope this accepts `DepEvents` by ref, because otherwise the same lifetime/retain concerns...